File: Laboratorio/Laboratorio04/java/Trapezoidal.java
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class Trapezoidal {
    public static double integrate(Function<Double, Double> function,
                                   double low,
                                   double sup,
                                   int n,
                                   int numThreads) throws InterruptedException {
        double height = (sup - low) / n;
        double firstAndLast = (function.apply(low) + function.apply(sup)) / 2.0;


        List<Thread> threads = new ArrayList<>();
        Box box = new Box();

        int chunk = n / numThreads;

        for (int i = 0; i < numThreads; i++) {
            int start = i * chunk + 1;
            int end = (i == numThreads - 1) ? n - 1 : (i + 1) * chunk;

            Thread worker = new Thread(() -> {
                double localSum = 0.0;
                for (int j = start; j <= end; j++) {
                    double xi = low + j * height;
                    localSum += function.apply(xi);
                }

                synchronized (box) {
                    box.value += localSum;
                }
            });

            threads.add(worker);
            worker.start();
        }

        for (Thread thread : threads) {
            thread.join();
        }

        return box.value * height + firstAndLast * height;
    }
}

File: Laboratorio/Laboratorio04/go/integrator/trapezoidal.go
package integrator

import (
	"sync"
)

func Trapezoidal(
	f func(float64) float64,
	low, sup float64,
	n, numWorkers int,
) float64 {

	height := (sup - low) / float64(n)
	firstAndLast := (f(low) + f(sup)) / 2.0

	var wg sync.WaitGroup
	var mu sync.Mutex

	sum := 0.0

	chunk := n / numWorkers

	for w := 0; w < numWorkers; w++ {
		start := w * chunk + 1
		end := (w + 1) * chunk
		if w == numWorkers-1 {
			end = n - 1
		}

		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			localSum := 0.0
			for i := start; i <= end; i++ {
				x := low + float64(i) * height
				localSum += f(x)
			}
			mu.Lock()
			sum += localSum
			mu.Unlock()
		}(start, end)
	}

	wg.Wait()
	return height * (firstAndLast + sum)
}